Batch async fetch shards data to reduce memory consumption.

howardhuanghua opened the pull request:
This commit is going to fix https://github.com/elastic/elasticsearch/issues/80694.
1. Queue shard async fetch requests and listeners.
2. Flush all the queued reqeusts after we collected all the node level async fetch reqeusts.
3. Split fetch responses to single `node-to-shard` and call the cached listeners after receiving node level fetch reqeusts.

Async shard fetch requests before/after optimization:
![image](https://user-images.githubusercontent.com/15325213/143686624-f09aa196-cb79-403f-8210-c78cb1ce6638.png)

![image](https://user-images.githubusercontent.com/15325213/144894205-2028a666-255f-4b38-b17c-5ddf06f1c533.png)

![image](https://user-images.githubusercontent.com/15325213/144894267-f96035dc-4634-4a54-9140-802439c0a304.png)

howardhuanghua committed:
Initial draft

howardhuanghua committed:
update draft

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -56,6 +56,7 @@
      */
     public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
         void list(ShardId shardId, @Nullable String customDataPath, DiscoveryNode[] nodes, ActionListener<NodesResponse> listener);
+        void flush();
```
Rather than introducing this method to the lister (and the corresponding flag passed in to `fetchData`) could we have the allocator directly indicate the end of an allocation round which triggers the flush.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -246,10 +258,132 @@ protected void reroute(ShardId shardId, String reason) {
+                        for (ShardId shardId : requestMap.keySet()) {
+                            ShardRequestInfo shardRequest = requestMap.get(shardId);
+                            shards.put(shardRequest.shardId(), shardRequest.getCustomDataPath());
+                            if (node.getVersion().before(Version.V_7_16_0)) {
```
The version in `master` is now 8.1.0; it's unlikely we'll backport this to an earlier version.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -246,10 +258,132 @@ protected void reroute(ShardId shardId, String reason) {
+                            };
+
+                            client.executeLocally(
+                                TransportNodesListGatewayStartedShards.TYPE,
```
I'm undecided about re-using the same action type for both kinds of request here. I think it'd be cleaner to introduce a new one (and to name it something better than `internal:gateway/local/started_shards`) given how big a difference in behaviour we are making.

howardhuanghua replied:
Hi @DaveCTurner , if we introduce a new action, then we need to refactor some logics in `GatewayAllocator`, like the follow structures, it seems that would a big change for the high level allocators. How do you think so?
https://github.com/elastic/elasticsearch/blob/2629c32efd53809f1792f093567e19cfcf44bf29/server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java#L57-L60

DaveCTurner replied:
I'm not sure this is true, I think we could keep pretty much the same interface from the point of view of `GatewayAllocator`. It should be possible to implement a batching `Lister` which reworks the batched responses into a `BaseNodesResponse<NodeGatewayStartedShards>`.

DaveCTurner commented on server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java:
```diff
@@ -97,37 +101,41 @@ protected NodeRequest newNodeRequest(Request request) {
-    protected NodeGatewayStartedShards nodeOperation(NodeRequest request, Task task) {
+    protected NodeGroupedGatewayStartedShards nodeOperation(NodeRequest request, Task task) {
+        NodeGroupedGatewayStartedShards groupedStartedShards = new NodeGroupedGatewayStartedShards(clusterService.localNode());
+        for (Map.Entry<ShardId, String> entry : request.getShards().entrySet()) {
```
When sending these requests per-shard we execute them in parallel across the `FETCH_SHARD_STARTED` threadpool. I think we should continue to parallelise them at the shard level like that.

DaveCTurner reviewed (commented):
This seems like the right sort of idea. I left some comments inline. I think we should also do this for replica allocations too.

howardhuanghua commented:
Thanks for the suggestion. I am going to complete the optimization.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

howardhuanghua committed:
Merge remote-tracking branch 'origin' into async_fetch_request

howardhuanghua committed:
Introduce new batch action for shard fetch

howardhuanghua committed:
fix compile issue

howardhuanghua committed:
revert shard store action

howardhuanghua committed:
remove unused func

howardhuanghua committed:
fix spotless issue

howardhuanghua committed:
add new action registion instance

howardhuanghua committed:
Merge remote-tracking branch 'origin' into async_fetch_request

howardhuanghua commented:
Hi @DaveCTurner I have finished the optimization.
1. Introduce new batch async fetch action. Use dynamic setting to control the old/new action switch.
2. Data nodes extract shard data in parallel.
3. Gateway allocator controls the pending requests flush operation.

Please help to review if you have free time. Thanks.

GrandFisher commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -283,11 +322,149 @@ private void listStartedShards(
+
+                synchronized (listenerMutex) {
+                    // queue child listeners
+                    List<ActionListener<BaseNodesResponse<NodeGatewayStartedShards>>> shardListeners = queuedListeners.computeIfAbsent(
```
Hi @howardhuanghua, the size of `queuedListeners` is O(shard) * O(dataNode),   and  it may takes up a lot of memory space When the cluster has more shards and data-nodes. Besides，many data fetch cache in asyncFetchData cannot be removed or cleared due to it need all the datanodes successful response.

howardhuanghua replied:
Hi @GrandFisher Thanks for the review.
1. Before receiving node level shards fetch results, the queued listeners are empty `ArrayList` elements. At the beginning of full cluster restart, node resource should be enough.
2. The queued node shard listeners would be cleared immediately once receiving the node level response.
3. Alternately, we could separate the shard requests in several of batches instead of flush all the shards in a big batch.

howardhuanghua replied:
After double checking, `queuedListerners` do have the risk to take up a lot of memory space.
I replaced `DiscoveryNode` with node id in single action listener. We will verify the memory consumption again. Thanks.

howardhuanghua committed:
merge master

howardhuanghua committed:
Replace discovery node with node id.

GrandFisher commented on server/src/main/java/org/elasticsearch/gateway/TransportNodesBatchListGatewayStartedShards.java:
```diff
@@ -0,0 +1,390 @@
+            final CountDownLatch latch = new CountDownLatch(request.getShards().size());
+            for (Map.Entry<ShardId, String> entry : request.getShards().entrySet()) {
+                threadPool.executor(ThreadPool.Names.FETCH_SHARD_STARTED).execute(() -> {
+                    batchStartedShards.addStartedShard(listStartedShard(entry.getKey(), entry.getValue()));
```
if `listStartedShard` throws a `ElasticsearchException`,  current thread `latch.countDown();` won't be executed , and the latch will never to be zero

howardhuanghua replied:
Good catch, during separate batch testing, I hit this issue. Now I am going to split several batchs to fetch huge number of shards. Like flush fetch request per 10k shards.

howardhuanghua committed:
Separate batch size.

howardhuanghua committed:
Merge remote-tracking branch 'origin' into async_fetch_request

howardhuanghua committed:
Fix style issue.

howardhuanghua committed:
fix spotless issue

howardhuanghua committed:
Fxi null pointer issue.

howardhuanghua commented:
Hi @GrandFisher I have updated the commit. Please help to review again when you have free time.
1. Removed the cached listeners, only cache fetching reqeusts.
2. Support separating batches.
3. Fix potential count down latch waiting forever issue in data node fetching process.

GrandFisher commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -283,11 +341,157 @@ private void listStartedShards(
+                    node,
+                    n -> new HashMap<>()
+                );
+                nodeLevelRequests.put(shardId, new ShardRequestInfo<>(shardId, customDataPath, listener));
```
@howardhuanghua  `ShardRequestInfo` need not  to create for each node, the listener in each `ShardRequestInfo` is the same listener.  This can reduce `ShardRequestInfo` Object number from O(shard) * O(node ) to O (shard)

howardhuanghua replied:
Thanks for the comment. I have updated commit to optimize this part.

howardhuanghua committed:
optimize shard request info

GrandFisher commented on server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -283,11 +341,155 @@ private void listStartedShards(
+                            NodeGatewayBatchStartedShards startedShardResponse = nodesBatchStartedShards.getNodes().get(0);
+                            List<NodeGatewayBatchStartedShard> startedShards = startedShardResponse.getStartedShards();
+                            assert startedShards.size() == curNodeRequests.size();
+                            for (NodeGatewayBatchStartedShard shard : startedShards) {
```
@howardhuanghua  What will happen if the node which the response from has shard throws  ElasticSearch Exception and didn't add to batchStartedShards at code `batchStartedShards.addStartedShard(listStartedShard(entry.getKey(), entry.getValue()));` in `TransportNodesBatchListGatewayStartedShards.java`.  
Can ignore them directly  ?

howardhuanghua replied:
Add fix to check the total target shards number.

howardhuanghua committed:
Fix conflict

howardhuanghua committed:
Merge remote-tracking branch 'origin' into async_fetch_request

howardhuanghua committed:
Fix shardId serialize issue and optimize failed shard case.

howardhuanghua committed:
support empty shard copy.

howardhuanghua committed:
revert unchange log

howardhuanghua committed:
Fix spotless.

howardhuanghua commented:
Hi @DaveCTurner , would you please help to review this PR again? Thank you.

howardhuanghua committed:
Add change log.

DaveCTurner commented:
Sorry for the long delay here @howardhuanghua, this is still (technically) on my list but it is not a priority at the moment and it's too large a change to fit its review in between other things. I think the general direction is good, but it'd help if we could break this down into several smaller (and easier-to-review) steps somehow.

howardhuanghua commented:
> it'd help if we could break this down into several smaller (and easier-to-review) steps somehow.

Thanks. I would try to break down into smaller steps. Any suggestion would be appreciated.

DaveCTurner commented:
> Thanks. I would try to break down into smaller steps. Any suggestion would be appreciated.

IMO the trickiest part is that today there's no well-defined "end" to the shard-by-shard fetching process, and there needs to be an end so that we know when to flush the last batch of requests. This PR adds this event (and a lot of other things besides). I think making that change in isolation would be worth trying as a first step. This change won't mean very much on its own because we won't actually be doing any flushing at the end, but still once we have this flushing event I believe it'll be easier to review the changes that adds batching on top of it.

Independently, I think you could detach `TransportIndicesShardStoresAction` from the `InternalAsyncFetch` mechanism. This action doesn't do any caching (and is hardly used anyway) so there's no need to involve it in these changes.

howardhuanghua commented:
Hi @DaveCTurner , in the follow logic of above PR, we first iterate all the unassigned primaries, each `allocator.allocateUnassigned` call would add single pending fetching shard, and after the while loop, we would call fush again to handle the rest of pending shards, this is the "end" that you mentioned above?

```
private void allocateExistingUnassignedShards(RoutingAllocation allocation) {
        allocation.routingNodes().unassigned().sort(PriorityComparator.getAllocationComparator(allocation)); // sort for priority ordering

        for (final ExistingShardsAllocator existingShardsAllocator : existingShardsAllocators.values()) {
            existingShardsAllocator.beforeAllocation(allocation);
        }

        GatewayAllocator gatewayAllocator = null;
        if (logger.isDebugEnabled()) {
            logger.debug("set batch fetch mode [{}] for routing allocation.", batchFetchShardEnable);
        }
        allocation.setBatchShardFetchMode(batchFetchShardEnable);

        final RoutingNodes.UnassignedShards.UnassignedIterator primaryIterator = allocation.routingNodes().unassigned().iterator();
        while (primaryIterator.hasNext()) {
            final ShardRouting shardRouting = primaryIterator.next();
            if (shardRouting.primary()) {
                ExistingShardsAllocator allocator = getAllocatorForShard(shardRouting, allocation);
                allocator.allocateUnassigned(shardRouting, allocation, primaryIterator);
                if (gatewayAllocator == null && allocator instanceof GatewayAllocator) {
                    gatewayAllocator = (GatewayAllocator) allocator;
                }

                if (gatewayAllocator != null
                    && gatewayAllocator.getPrimaryPendingFetchShardCount() > 0
                    && gatewayAllocator.getPrimaryPendingFetchShardCount() % batchFetchShardStepSize == 0) {
                    gatewayAllocator.flushPendingPrimaryFetchRequests(batchFetchShardStepSize);
                }
            }
        }

        // flush the rest primaries
        if (gatewayAllocator != null) {
            gatewayAllocator.flushPendingPrimaryFetchRequests(batchFetchShardStepSize);
        }
```
Original code:
https://github.com/TencentCloudES/elasticsearch/blob/e415426c511088ca5b0d4e86b205a6ab2b025648/server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java#L592-L595

elasticsearchmachine commented:
Pinging @elastic/es-distributed-obsolete (Team:Distributed (Obsolete))

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)
